axlearn/open_api/metrics/tool_use_execution_utils.py

```diff
@@ -0,0 +1,334 @@
+    """Performs a lenient string transformation."""
+    normalized_str_value = str_value.lower()
+    normalized_str_value = normalized_str_value.translate(str.maketrans("", "", string.punctuation))
+    words = normalized_str_value.split()
+    start_index = next(
+        (i for i in range(len(words)) if words[i] not in _STOP_WORDS),
+        len(words) - 1 if len(words) > 0 and words[-1] not in _STOP_WORDS else len(words),
+    )
+    end_index = next(
+        (
+            i + 1
+            for i in reversed(range(start_index + 1, len(words)))
+            if words[i] not in _STOP_WORDS
+        ),
+        start_index + 1,
+    )
+
+    words = words[start_index:end_index]
+    normalized_str_value = " ".join(words)
+    return normalized_str_value
+
+
+def _word_set(input_str: str) -> set[str]:
+    """Returns a set of words splitting by whitespace, newline and tab,
+    and skipping empty strings from the set."""
+    return {s for s in re.split(r"\s+", input_str) if s}
+
+
+def _match_strings_bag_of_words(*, pred_str: str, target_str: str, threshold: float = 1.0) -> bool:
+    """Match strings using a bag of words approach.
+
+    Args:
+        pred_str: The predicted argument string.
+        target_str: The target argument string.
+        threshold:  Thresold to be compared with the ratio (# unique common words) / (# unique
+            pred_str words). The predicted string is considered to match the target if the ratio
+            is higher or equal to this threshold.
+
+    Returns:
+        True if recall is higher or equal to threshold; otherwise return False.
+    """
+    if pred_str == target_str:
+        return True
+
+    if threshold <= 0:
+        raise ValueError(
+            f"Bag of words string matching threshold must be above 0, but is {threshold}."
+        )
+    pred_word_set = _word_set(pred_str)
+    target_word_set = _word_set(target_str)
+
+    if len(pred_word_set) == 0 and len(target_word_set) == 0:
+        return True
+    if len(target_word_set) == 0:
+        return False
+
+    common_words = target_word_set.intersection(pred_word_set)
+    ratio = len(common_words) / len(target_word_set)
+    return ratio >= threshold
+
+
+def _is_arg_value_equal(
+    *,
+    pred_arg: ValueOrListOf,
+    target_arg: ValueOrListOf,
+    match_type: ArgumentMatchType,
+) -> bool:
+    """Checks if the predicted and target arguments are equal under different checks."""
+    if match_type == ArgumentMatchType.STRICT:
+        return pred_arg == target_arg
+
+    if (
+        isinstance(pred_arg, list)
+        and isinstance(target_arg, list)
+        and len(pred_arg) == len(target_arg)
+    ):
+        return all(
+            _is_arg_value_equal(
+                pred_arg=el_pred,
+                target_arg=el_target,
+                match_type=match_type,
+            )
+            for el_pred, el_target in zip(pred_arg, target_arg)
+        )
+    # Only handling string payloads for lenient evaluation
```

Comment:
```suggestion
    # Only handling string payloads for lenient evaluation.
```
(A gentle reminder to apply review comments throughout the PR.)

Reply:
Done